test/e2e/test/kibana/steps_mutation.go

```diff
@@ -5,20 +5,72 @@
 package kibana
 
 import (
+	"fmt"
 	"testing"
 
 	"github.com/elastic/cloud-on-k8s/pkg/apis/kibana/v1beta1"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/common/hash"
 	"github.com/elastic/cloud-on-k8s/pkg/utils/k8s"
 	"github.com/elastic/cloud-on-k8s/test/e2e/test"
 	"github.com/stretchr/testify/require"
+	corev1 "k8s.io/api/core/v1"
 )
 
 func (b Builder) MutationTestSteps(k *test.K8sClient) test.StepList {
-	return b.UpgradeTestSteps(k).
+	return b.AnnotatePodsWithBuilderHash(k).
+		WithSteps(b.UpgradeTestSteps(k)).
 		WithSteps(b.CheckK8sTestSteps(k)).
 		WithSteps(b.CheckStackTestSteps(k))
 }
 
+func (b Builder) AnnotatePodsWithBuilderHash(k *test.K8sClient) test.StepList {
+	return []test.Step{
+		{
+			Name: "Annotate Pods with a hash of their Builder spec",
+			Test: test.Eventually(func() error {
+				if b.MutatedFrom == nil {
+					panic("builder has to have MutatedFrom set if it's a mutation builder")
+				}
+
+				var pods corev1.PodList
+				if err := k.Client.List(&pods, test.KibanaPodListOptions(b.Kibana.Namespace, b.Kibana.Name)...); err != nil {
+					return err
+				}
+
+				for _, pod := range pods.Items {
+					if pod.Annotations == nil {
+						pod.Annotations = make(map[string]string)
+					}
+					pod.Annotations[BuilderHashAnnotation] = hash.HashObject(b.MutatedFrom.Kibana.Spec)
+					if err := k.Client.Update(&pod); err != nil {
+						// may error out with a conflict if concurrently updated by the operator,
+						// which is why we retry with `test.Eventually`
+						return err
+					}
+				}
+				return nil
+			}),
+		},
+		// make sure this is propagated to the local cache so next test steps can expect annotated pods
```

Comment:
I think I wrote this for the elasticsearch side of things but it's probably useless: the k8s client we use in E2E tests does not rely on a cached reader. Which in the context of E2E tests makes sense, I think (the performance hit is fine?).
Could you double-check I am not mistaken here and remove this block (also on ES side) if it makes sense? 🙏